src/main/java/javaslang/collection/Traversable.java

```diff
@@ -553,6 +540,77 @@ default T fold(T zero, BiFunction<? super T, ? super T, ? extends T> op) {
     Option<T> headOption();
 
     /**
+     * Returns the index of the first occurrence of the given element or -1 if this does not contain the given element.
+     *
+     * @param element an element
+     * @return the index of the first occurrence of the given element
+     */
+    default int indexOf(T element) {
+        return indexOf(element, 0);
+    }
+
+    /**
+     * Returns the index of the first occurrence of the given element after or at some start index
+     * or -1 if this does not contain the given element.
+     *
+     * @param element an element
+     * @param from start index
+     * @return the index of the first occurrence of the given element
+     */
+    int indexOf(T element, int from);
+
+    /**
+     * Finds first index where this sequence contains a given sequence as a slice.
+     * <p>
+     * Note: may not terminate for infinite-sized collections.
+     * @param that the sequence to test
+     * @return the first index such that the elements of this sequence starting at this index match
+     * the elements of sequence that, or -1 of no such subsequence exists.
+     * @throws NullPointerException if {@code that} is null.
+     */
+    default int indexOfSlice(Iterable<? extends T> that) {
+        Objects.requireNonNull(that, "that is null");
+        return indexOfSlice(that, 0);
+    }
+
+    /**
+     * Finds first index after or at a start index where this sequence contains a given sequence as a slice.
+     * <p>
+     * Note: may not terminate for infinite-sized collections.
+     * @param that the sequence to test
+     * @param from the start index
+     * @return the first index >= from such that the elements of this sequence starting at this index match
+     * the elements of sequence that, or -1 of no such subsequence exists.
+     * @throws NullPointerException if {@code that} is null.
+     */
+    default int indexOfSlice(Iterable<? extends T> that, int from) {
```

Comment:
@szarnekow Good Morning! Did you mean

> `Iterable<?>` rather than `Iterable<? extends T>`

or

> `Iterable<? super T>` rather than `Iterable<? extends T>`

We need the type-safety in general I think. It would be possible to first 'downgrade' the traversable, e.g.

```java
traversable                     // is a Traversable<String>
        .as(CharSequence.class) // returns a Traversable<CharSequence>
        .indexOfSlice(that, 0)
```

Unfortunately it is not possible in Java to define a generic lower bound `<U super T>` (which I do not understand but which will have to do with the compiler algorithm to that finds correct bounds):

```java
<U super T> Traversable<U> as(Class<U> type) {
    return (Traversable<U>) this;
}
```

That means that `Iterable<?>` would be the only solution to allow lower bounds (`Iterable<? super T> that`) and upper bounds (`Iterable<? extends T> that`).

My impression also is, that `Iterable<?>` will be error-prone.